server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -414,13 +418,17 @@ final boolean syncUpTo(long offset) throws IOException {
                     // now do the actual fsync outside of the synchronized block such that
                     // we can continue writing to the buffer etc.
                     try {
-                        channel.force(false);
+                        if (lastSyncedCheckpoint.offset != checkpointToSync.offset) {
+                            channel.force(false);
+                        }
                         writeCheckpoint(checkpointChannel, checkpointPath, checkpointToSync);
```

Comment:
This seems wrong, it leaves us with a (synced) checkpoint that refers to data that isn't synced and might vanish if, say, you pull the power out after executing this line.

Reply:
Ah wait I see, if the `offset` didn't change then we didn't actually write anything. That deserves an assertion for sure, if only for documentation.